### DISCUSSION TITLE
how to get self-position in localization mode

### DISCUSSION THREAD

**Discussion by masaki12-s** [introlab/rtabmap#970] [Q&A]
I am using RTABMAP on Windows 10 for localization.
I want to get the coordinates within a room.
After generating a 3D map in "Mapping" mode, I would like to get the self-position in "Localization" mode.
Where is the estimated self-position stored in "Localization" mode?

**Comment by matlabbe** [comment]
What do you mean by "how to get"? programmatically? In localization mode, you should see the pose of the virtual camera frame in 3D map view matching where it is localized. In code, the localized pose is determined by latest odometry pose and the map correction. 

Here is where the camera frame of 2d view is updated on odometry event:

https://github.com/introlab/rtabmap/blob/2acd8f82b58f07fd1e224e79d560638c24c07234/guilib/src/MainWindow.cpp#L1588

The `_odometryCorrection` is updated when receiving a new rtabmap event:
https://github.com/introlab/rtabmap/blob/2acd8f82b58f07fd1e224e79d560638c24c07234/guilib/src/MainWindow.cpp#L2394

**Reply by masaki12-s** [reply]
I would like to get the coordinates on the windows GUI.
" Mapping" mode is mapping with self location and the trajectory is drawn as a blue line. The coordinates can be obtained from "poses.txt".
however, in "localization" mode, the trajectory is drawn as a pink line, which disappears over time. Also, it cannot be retrieved from "poses.txt". I would like to get the coordinates of this pink line.

**Reply by takanotume24** [reply]
@matlabbe Hello, thank you for your detailed answer. I am a participant in the same project as masaki12-s. Thank you very much for your detailed answer.

What we wanted to know is how to get the self-position in localized mode programmatically, as you said, it would be better to get the pose of the virtual camera frame, we are using RTAB-Map for the purpose of getting the self-position after generating the 3D map. I will try to get it from the C++ code you taught me. Thank you for showing us the relevant code.

Also, as an addendum to masaki12-s question, is there a mode implemented in the Windows version of the GUI software to get the pose of the virtual camera frame with timestamp?

**Reply by matlabbe** [reply]
The pose is published by odometry event. An example of handling the odometry pose and transform it in current map frame can be found here: 
https://github.com/introlab/rtabmap/blob/0f221ba3cd9913b68ec508af0d914e3f49860430/guilib/src/MainWindow.cpp#L1590

For simpler example, a similar approach is done in the c++ example:
https://github.com/introlab/rtabmap/blob/0f221ba3cd9913b68ec508af0d914e3f49860430/examples/RGBDMapping/MapBuilder.h#L156
